doc/design.md

```diff
@@ -0,0 +1,1268 @@
+`numpy`, `pandas`, and `xarray` also adopt these semantics.
+The disadvantage is that operations cannot be done in-place without risking breaking all other instances.
+In-place operations would furthermore not be thread-safe in many cases.
+
+A compromise between the two is the copy-on-write mechanism that has since long been adopted in Mantid, e.g., in `DataObjects::Workspace2D` for the histogram data and for `Beamline::DetectorInfo`.
+Qt uses a slightly different implementation referred to as [implicit sharing](http://doc.qt.io/qt-5/implicit-sharing.html) for many of their basic data types.
+This combines the advantages of easy-to-reason-about value semantics with no-overhead copying or partial modification.
+
+The cost is *invalidation of all iterators, references, or views* to the underlying data if *any* non-`const` method is called or if a copy is made.
+This is also made explicit in the [corresponding Qt](http://doc.qt.io/qt-5/containers.html#the-iterator-classes) [documentation](http://doc.qt.io/qt-5/containers.html#implicit-sharing-iterator-problem).
+The effect is most visible in multi-threaded environments, but can be demonstrated even without:
+
+```cpp
+#include "MantidKernel/cow_ptr.h"
+
+Kernel::cow_ptr<std::vector<double>> vec;
+vec.access().resize(1000);
+auto copy(vec);
+
+auto const_it = vec->begin();
+vec.access()[0] = 1.0;
+// const_it is now pointing to data owned exclusively by copy.
+
+copy.reset(); // This could be DeleteWorkspace in another (GUI) thread.
+// const_it is now pointing to free'ed memory.
+```
+
+Note that this trap is absolutely not obvious since only the middle section of the above block would typically be part of the local source code.
+For example, many algorithms support `OutputWorkspace == InputWorkspace` and may call something like
+```cpp
+// This is illegal, do not do this, unless you verified that outputWS != inputWS.
+outputWS.mutableY(0) = inputWS.y(0);
+```
+
+If `outputWS == inputWS` and the reference count for the Y data is not 1 then `inputWS.y(0)` references data that is now owned by other workspace instances or by other histograms in the same workspace.
+That is, it can be arbitrarily modified or even get deallocated.
+
+Note that assignment (`operator=`) [is not a sequence point in C++](https://en.cppreference.com/w/cpp/language/eval_order), i.e., this problem does indeed occur even if neither the LHS nor the RHS in above example is explicitly stored as a reference.
+
+
+##### Conclusion and rationale
+
+- Neither of the extreme cases 1.) or 2.) seems to be a viable alternative.
+- Use a copy-on-write mechanism that is hidden in the API and simply handled internally in `Variable`:
+  A copy-on-write pointer holds a `VariableConcept`.
+  To maintain sharing we only require code to be `const`-correct.
+  *Question: Cannot use `const` in Python, need methods/tags with different names that explicitly request mutable or immutable access. What is the best naming convention?*
```

Comment:
The std library added cbegin to refer to a const_iterator. Perhaps just prefix the const methods with c? Or is this too short?